Remotion/Data/DomainObjects/Persistence/ReadOnlyStorageProviderManager.cs

```diff
@@ -0,0 +1,129 @@
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Remotion.Data.DomainObjects.Persistence.Configuration;
+using Remotion.Data.DomainObjects.Tracing;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Persistence
+{
+  public class ReadOnlyStorageProviderManager : IReadOnlyStorageProviderManager
+  {
+    private bool _disposed;
+    private Dictionary<string, IReadOnlyStorageProvider>? _storageProviders;
+    private readonly IPersistenceExtension _persistenceExtension;
+    private readonly IStorageSettings _storageSettings;
+
+    public ReadOnlyStorageProviderManager (IPersistenceExtension persistenceExtension, IStorageSettings storageSettings)
+    {
+      ArgumentUtility.CheckNotNull("persistenceExtension", persistenceExtension);
+      ArgumentUtility.CheckNotNull("storageSettings", storageSettings);
+
+      _storageProviders = new Dictionary<string, IReadOnlyStorageProvider>();
+      _persistenceExtension = persistenceExtension;
+      _storageSettings = storageSettings;
+    }
+
+    #region IDisposable Members
+
+    public void Dispose ()
+    {
+      if (!_disposed)
+      {
+        foreach (var keyValue in _storageProviders!)
+        {
+          keyValue.Value.Dispose();
+        }
+
+        _storageProviders = null;
+        _disposed = true;
+      }
+    }
+
+    #endregion
+
+    public IReadOnlyStorageProvider GetMandatory (string storageProviderID)
+    {
+      CheckDisposed();
+      ArgumentUtility.CheckNotNullOrEmpty("storageProviderID", storageProviderID);
+
+      if (_storageProviders.TryGetValue(storageProviderID, out var storageProvider))
+        return storageProvider!;
+
+      var providerDefinition = _storageSettings.GetStorageProviderDefinition(storageProviderID);
+      var provider = providerDefinition.Factory.CreateStorageProvider(providerDefinition, _persistenceExtension);
+
+      if (provider == null)
+        throw CreatePersistenceException("Storage provider '{0}' could not be created.", storageProviderID);
+
+      _storageProviders.Add(storageProviderID, provider);
+
+      return provider;
+    }
+
+    public IReadOnlyStorageProvider GetMandatory (StorageProviderDefinition providerDefinition)
+    {
+      CheckDisposed();
+      ArgumentUtility.CheckNotNull("providerDefinition", providerDefinition);
+
+#if DEBUG
+      if (providerDefinition != _storageSettings.GetStorageProviderDefinition(providerDefinition.Name))
+      {
+        throw new InvalidOperationException(
+            $"Supplied provider definition '{providerDefinition.Name}' does not match the provider definition with the same name in the IStorageSettings object.");
+      }
+#endif
+
+      if (_storageProviders.TryGetValue(providerDefinition.Name, out var storageProvider))
+        return storageProvider!;
```

Comment:
`storageProvider` cannot be `null`, because we have a `Dictionary<string, IReadOnlyStorageProvider>`, where the value type is non-nullable, so the `!` does not do anything.